ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/DiffSeq.hs

```diff
@@ -0,0 +1,354 @@
+   means of the @'invert'@ operation that the @'Group'@ type class requires.
+   Namely, if either the left or right part of the split is small in comparison
+   with the input sequence, then we can subtract the diffs in the smaller part
+   from the root measure of the input to (quickly) compute the root measure of
+   the /other/ part of the split. This is much faster than computing the root
+   measures from scratch by doing a linear-time pass over the elements of the
+   split parts, or a logarithmic-time pass over intermediate sums of diffs in
+   case we store cumulative diffs in the nodes of the finger tree.
+
+   === Example of fast splits
+
+   As an analogy, consider this example: we have a sequence of consecutive
+   integer numbers @xs = [1..n]@ where @n@ is large, and we define the root
+   measure of the sequence to be the total sum of these numbers, @rmxs = sum
+   [1..n]@ (we assume @rmxs@ is fully evaluated). Say we split this sequence of
+   integer numbers at the index @2@, then we get /left/ and /right/ parts of the
+   split @ys@ and @zs respectively.
+
+   > splitAt 2 xs = (ys, zs) = ([1..2], [3..n])
+
+   How should we compute we the root measure @rmys@ of @ys@? Since @ys@ is
+   small, we can just compute @rmys = sum [1..2]@. How should we compute the
+   root measure @rmzs@ of @zs@? We should not compute @rmzs = sum [3..n]@ in this
+   case, since @n@ is large. Instead, we compute @rmzs = rmxs - rmys@, which
+   evaluates to its result in time that is linear in the length of @ys@, in this
+   case @O(2)@.
+
+   === Why not store sums of diffs in the internal measure instead of the root
+   measure?
+
+   We could also have used the interal measure of the strict finger tree to
+   store intermediate sums of diffs for all subtrees of the node. The subtree
+   rooted at the root of the tree would then store the total sum of diffs.
+   However, we would have now to recompute a possibly logarithmic number of sums
+   of diffs when we split or extend the sequence. Given that in @consensus@ we
+   use the total sum of diffs nearly as often as we split or extend the diff
+   sequence, this proved to be too costly. The single-instance root measure
+   reduces the overhead of this "caching" of intermediate sums of diffs by only
+   using a single total sum of diffs, though augmented with an @'invert'@
+   operation to facilitate computing updated root measures.
+
+   === Note on the use of @l@ and @r@ infixes in function names
+
+   Functions like @'splitl'@, @'splitlAt'@ and @'splitrAtFromEnd'@ use infix @l@
+   and @r@ characters to indicate that its time complexity is dependent on
+   either the length of the left part or right part of the split respectively.
+   This means that if we expect the left part of a split to be small relative to
+   the length of the input, then we should use an @l@-variant. If the reverse is
+   true, then we should use an @r@-variant.
```

Comment:
Are there parts of the code that consider the length of the sequence to use either an `l` or an `r` variant?

Reply:
Not currently, but it would not surprise if that at some point we would move to using some heuristic to make the decision, instead of hard-coding which variant we use